assets/js/wc-gateway-ppec-generate-cart.js

```diff
@@ -30,52 +30,66 @@
 			$( '#woo_pp_ec_button_product > *' ).css( 'pointer-events', 'none' );
 		} );
 
+	// True if the product is simple or the user selected a valid variation. False on variable product without a valid variation selected
+	var variation_valid = true;
+
+	// True if all the fields of the product form are valid. False otherwise
+	var fields_valid = true;
+
+	// True if there's an automatic form validation being performed (not user-initiated), so the user must not be alerted. False otherwise
+	var silent_validation = false;
+
+	var update_button = function() {
+		$( '#woo_pp_ec_button_product' ).trigger( ( variation_valid && fields_valid ) ? 'enable' : 'disable' );
+	};
+
+	var validate_form = function() {
+		silent_validation = true;
+		fields_valid = form.get( 0 ).checkValidity();
+		silent_validation = false;
+		update_button();
+	};
+
 	// It's a variations form, button availability should depend on its events
 	if ( $( '.variations_form' ).length ) {
-		$( '#woo_pp_ec_button_product' ).trigger( 'disable' );
+		variation_valid = false;
 
 		$( '.variations_form' )
 		.on( 'show_variation', function( event, form, purchasable ) {
-			$( '#woo_pp_ec_button_product' ).trigger( purchasable ? 'enable' : 'disable' );
+			variation_valid = purchasable;
+			update_button();
 		} )
 		.on( 'hide_variation', function() {
-			$( '#woo_pp_ec_button_product' ).trigger( 'disable' );
+			variation_valid = false;
+			update_button();
 		} );
 	}
 
-	var get_attributes = function() {
-		var select = $( '.variations_form' ).find( '.variations select' ),
-			data   = {},
-			count  = 0,
-			chosen = 0;
-
-		select.each( function() {
-			var attribute_name = $( this ).data( 'attribute_name' ) || $( this ).attr( 'name' );
-			var value	  = $( this ).val() || '';
-
-			if ( value.length > 0 ) {
-				chosen++;
-			}
-
-			count++;
-			data[ attribute_name ] = value;
-		} );
-
-		return {
-			'count'      : count,
-			'chosenCount': chosen,
-			'data'       : data
-		};
-	};
+	// Disable the button if there are invalid fields in the product page (like required fields from Product Addons)
+	var form = $( 'form.cart' );
```

Comment:
What do you think of moving the `form` definition closer to the top as well? It's being used in `validate_form()` above so it would be greater for the reader to see it earlier.

Reply:
Yeah, done 👍